standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+---
+title: WAKU2-RLN-CONTRACT
+name: Waku2 RLN Contract Specification
+category: Standards Track
+tags: [waku/core-protocol]
+editor: Sergei Tikhomirov <sergei@status.im>
+contributors:
+---
+
+## Abstract
+
+This document describes membership management for the RLN smart contract, in particular:
+- membership-related contract functionality;
+- suggested parameters valued for the initial mainnet deployment;
+- contract governance and upgradability.
+
+This document currently only considers membership-related functionality.
+It might later evolve into a full-fledged contract specification.
+
+RLN is only deployed on Sepolia testnet ([source code](https://github.com/waku-org/waku-rlnv2-contract/blob/main/src/WakuRlnV2.sol)) as of August 2024.
+This document aims to outline the path to its mainnet deployment.
+
+## Background
+
+Rate-Limiting Nullifier (RLN) is a ZK-based gadget used for privacy-preserving rate limiting in Waku.
+The RLN smart contract is the core element of RLN architecture.
+The smart contract stores the RLN tree that contains all currently existing memberships.
+Users interact with the contract to manage their memberships
+and to get the necessary data for proof generation and verification.
+
+Sending messages is handled by Waku Relay nodes.
+To send a message, the sender MUST attach a ZK-proof to the message.
+A Relay node MUST relay a message unless:
+- the message belongs to any epoch other than the current one; or
+- the user has exceed their allowed rate limit for the current epoch; or
+- the ZK-proof fails to prove that the sender holds a membership.
+
+The full specification of Relay is out of scope for this document.
+
+## Contract overview
+
+The contract MUST provide the following functionalities:
+- register a membership;
+- extend a membership;
+- withdraw a deposit.
+
+Contract parameters and their RECOMMENDED values for the initial mainnet deployment are as follows:
+
+| Parameter                                   | Symbol    | Value   | Units                              |
+| ------------------------------------------- | --------- | ------- | ---------------------------------- |
+| Epoch length                                | `epoch`   | `10`    | minutes                            |
+| Maximum total rate limit of all memberships | `R_{max}` | `20000` | messages per `epoch`               |
+| Minimal rate limit of one membership        | `r_{min}` | `20`    | messages per `epoch`               |
```

Comment:
Do we set the maximum somewhere? Presumably we want to prevent one membership from dominating the total available rate?

Reply:
My initial reaction was - sure, let's introduce a maximum. But thinking about it a bit more, I'm not so certain.

An attacher who wants can easily register many memberships to dominate the total rate. We don't have an identity mechanism to stop it. Even if we limit the number of memberships per Ethereum address (currently we explicitly allow it), the attacker would create as many addresses as needed.

Should we specify a restriction that we can't enforce? Or can we?

Reply:
Another point on this is that the max rate limit is an issue for project integrating Waku. By not setting one, we give more flexibility for projects to discover and play with right parameter.
The only issue being all project overshooting the rate limit and depleting it without the actual usage being anywhere near the rate limit.

Which then is an argument in favour of a max, to ensure projects/application developers don't overshoot. Especially consider how cheap a membership is.

Reply:
Based on our discussion elsewhere, I've set the max rate limit for one membership to 600 msg / epoch (= 1 msg / s at the recommended epoch length).